### PR TITLE
implementing `PREPARE`, `EXECUTE`, and `DEALLOCATE`

### PR DESCRIPTION
**Pull request by JCOR11599**
fx for: https://github.com/dolthub/dolt/issues/4452

### CODE REVIEW

File: enginetest/memory_engine_test.go [hunk 1]
```diff
@@ -183,26 +183,31 @@ func TestSingleQueryPrepared(t *testing.T) {
 
 // Convenience test for debugging a single query. Unskip and set to the desired query.
 func TestSingleScript(t *testing.T) {
-	t.Skip()
```
Comment (fulghum) [inline enginetest/memory_engine_test.go]: Go ahead and tidy this up so it's nice and clean and ready for the next person who needs to use it (including removing all the commented out code below)

File: sql/errors.go [hunk 2]
```diff
@@ -247,6 +247,9 @@ var (
 	// ErrUnboundPreparedStatementVariable is returned when a query is executed without a binding for one its variables.
 	ErrUnboundPreparedStatementVariable = errors.NewKind(`unbound variable "%s" in query`)
 
+	// ErrUnknownPreparedStatement is returned when an unknown query is executed.
+	ErrUnknownPreparedStatement = errors.NewKind(`Unknown prepared statement handler (%s) given to EXECUTE`)
```
Comment (fulghum) [inline sql/errors.go:251]:
```suggestion
	ErrUnknownPreparedStatement = errors.NewKind(`Unknown prepared statement (%s) given to EXECUTE`)
```

File: sql/parse/parse.go [hunk 3]
```diff
@@ -273,6 +273,16 @@ func convert(ctx *sql.Context, stmt sqlparser.Statement, query string) (sql.Node
 		return plan.NewShowPrivileges(), nil
 	case *sqlparser.Flush:
 		return convertFlush(ctx, n)
+	case *sqlparser.Prepare:
+		newStmt, err := sqlparser.Parse(n.Expr)
+		if err != nil {
+			return nil, err
+		}
+		return convertPrepare(ctx, newStmt, n)
```
Comment (fulghum) [inline sql/parse/parse.go:281]: (minor) lines 277-280 feel pretty coupled with `convertPrepare`, so you might try pushing the responsibility for parsing the inner statement into the `convertPrepare` function and keep this large convert function a little simpler.

File: sql/plan/prepare.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
+// Copyright 2020-2022 Dolthub, Inc.
```
Comment (fulghum) [inline sql/plan/prepare.go:1]: (super minor) For new source files, just start with the current year for the copyright header comment. 

```suggestion
// Copyright 2022 Dolthub, Inc.
```

File: sql/plan/prepare.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// PrepareQuery is a node that prepares the query
+type PrepareQuery struct {
+	Name  string
+	Child sql.Node
+}
+
```
Comment (fulghum) [inline sql/plan/prepare.go:28]: Good practice to always add the `var _ sql.Node = (*PrepareQuery)(nil)` lines to get type checking that our implementations always fully implement any interfaces we want them to. It's a great safeguard to have in the future when we inevitably need to change those interfaces.

File: engine.go [hunk 6]
```diff
@@ -334,6 +351,73 @@ func (e *Engine) analyzeQuery(ctx *sql.Context, query string, parsed sql.Node, b
 		return nil, err
 	}
 
+	switch n := parsed.(type) {
```
Comment (fulghum) [inline engine.go:354]: I haven't thought through this fully yet... so I might be missing something... but I'm wondering if there is another way to execute these statements that is more similar to how we execute other statements. It seems inconsistent that these statements are "executed" directly inside `engine.go`. I'm wondering if that creates other edge cases. 

I see that they need to access the engine to call some functions, but I'm wondering if you had the `PreparedStatementCache` type I mentioned in a previous comment that could be set in these types, if that would decouple them.

File: enginetest/queries/script_queries.go [hunk 7]
```diff
@@ -3392,6 +3392,141 @@ var CreateCheckConstraintsScripts = []ScriptTest{
...
+				},
+			},
+			{
+				Query:       "execute s using @a",
+				ExpectedErr: sql.ErrUnknownPreparedStatement,
+			},
+		},
+	},
+	{
+		Name: "prepare insert",
```
Comment (fulghum) [inline enginetest/queries/script_queries.go:3487]: Would be nice to add a more complex test, with some subqueries and maybe a derived table or something in it, just to test something a little more complex.

**Review by fulghum — COMMENTED** [review]
Looking pretty good. I think it's worth going a little further to tidy things up though.

File: sql/errors.go (hunk 2 above)
Comment (JCOR11599) [reply sql/errors.go:251]: I just copied MySQL's error for this
```sql
mysql> execute nonexistent;
ERROR: 1243: Unknown prepared statement handler (nonexistent) given to EXECUTE
```

File: engine.go (hunk 6 above)
Comment (JCOR11599) [reply engine.go:367]: I do agree it's a little strange to have these in the `engine.go` file, but I couldn't think of another place to put it.
I'm not quite sure what happens when you prepare a prepare statement either, but I'm assuming not great things.

Comment (fulghum) [reply engine.go:376]: Good edge case with preparing a `prepare` statement! It looks like MySQL's grammar doesn't allow this, so it's always a syntax error. Would definitely be a good test case to cover.

File: enginetest/queries/script_queries.go (hunk 7 above)
Comment (fulghum) [reply enginetest/queries/script_queries.go:3487]: I see some error case tests for `execute`; it would be good to add some for `prepare` with an invalid statement and maybe a quick one for `deallocate` with an unknown statement id.

File: engine.go [hunk 8]
```diff
@@ -55,6 +55,65 @@ type TemporaryUser struct {
 	Password string
 }
 
+// PreparedDataCache manages all the prepared data for every session for every query for an engine
+type PreparedDataCache struct {
```
Comment (fulghum) [inline engine.go:59]: This is beautiful! It encapsulates this logic cleanly, shrinks the locking granularity, and makes this code a lot easier to read. Nice work! 👏

File: engine.go (hunk 6 above)
Comment (fulghum) [reply engine.go:376]: Now that you have `PreparedDataCache`... it seems like you're closer to being able to pull this out of `engine.go`... if you could set the `PreparedDataCache` instance on these plan nodes, then they could run their logic to set/delete in the cache as part of their rowIter func. Seems like we would need another analyzer rule to catch these nodes and run the Analyzer calls to `PrepareQuery` and `AnalyzedPrepared` though. 

@max-hoffman  – what do you think? any better ideas for us here?

File: engine.go [hunk 9]
```diff
@@ -334,6 +373,51 @@ func (e *Engine) analyzeQuery(ctx *sql.Context, query string, parsed sql.Node, b
...
+		analyzed, _, err = e.Analyzer.AnalyzePrepared(ctx, parsed, nil)
+		if err != nil {
+			return nil, err
+		}
+		return analyzed, nil
+	case *plan.DeallocateQuery:
+		if _, ok := e.PreparedDataCache.GetCachedStmt(ctx.Session.ID(), n.Name); !ok {
+			return nil, sql.ErrUnknownPreparedStatement.New(n.Name)
+		}
+		e.PreparedDataCache.UncacheStmt(ctx.Session.ID(), n.Name)
```
Comment (max-hoffman) [inline engine.go:418]: should this return on completion?

File: engine.go [hunk 10]
```diff
@@ -334,6 +373,51 @@ func (e *Engine) analyzeQuery(ctx *sql.Context, query string, parsed sql.Node, b
 		return nil, err
 	}
 
+	switch n := parsed.(type) {
+	case *plan.PrepareQuery:
+		analyzedChild, err := e.Analyzer.PrepareQuery(ctx, n.Child, nil)
+		if err != nil {
+			return nil, err
+		}
+		e.PreparedDataCache.CacheStmt(ctx.Session.ID(), n.Name, analyzedChild)
```
Comment (max-hoffman) [inline engine.go:382]: same here, return?

File: engine.go (hunk 9 above)
Comment (JCOR11599) [reply engine.go:418]: Should the function `UncacheStmt` return something? or should the switch statement return something?
I could return in each case of the switch, since the analyzer has no rules for `PrepareQuery`, `ExecuteQuery`, or `DeallocateQuery`, so there's no reason to analyze them.

File: engine.go (hunk 6 above)
Comment (max-hoffman) [reply engine.go:376]: i think the answer depends on whether PREPARE and DEALLOCATE are exempted from the transactional lifecycle. for example, if you start a transaction with an already prepared query, unbind a prepared statement, and then rollback the unbind, does MySQL restore the prepared query? or did the unbind not respect transaction boundaries? the reverse -- prepare a statement, and then rollback -- does the session still see the cached plan?

File: engine.go (hunk 9 above)
Comment (fulghum) [reply engine.go:418]: Yup; I read that as "returning in each case of the switch" (like you're already doing for ExecuteQuery)

**Review by fulghum — APPROVED** [review]
Thanks for iterating on this one! I think your changes really improved this code a ton.
